src/fides/api/ops/models/privacy_experience.py

```diff
@@ -165,26 +297,302 @@ def update(self, db: Session, *, data: dict[str, Any]) -> PrivacyExperience:
+            )
+            .first()
+        )
+
+        return overlay_experience, privacy_center_experience
+
+    def unlink_privacy_experience_config(self, db: Session) -> PrivacyExperience:
+        """Remove config from experience"""
+        return self.update(
+            db,
+            data={
+                "experience_config_id": None,
+                "experience_config_history_id": None,
+            },
+        )
 
 
 class PrivacyExperienceHistory(PrivacyExperienceBase, Base):
-    """Stores historical records of Privacy Experiences for Consent Reporting"""
+    """Stores the history of a privacy experience for a given region for Consent Reporting"""
 
     version = Column(Float, nullable=False, default=1.0)
-    privacy_experience_template_id = Column(
-        String, ForeignKey(PrivacyExperienceTemplate.id_field_path), nullable=True
-    )
     privacy_experience_id = Column(
-        String, ForeignKey(PrivacyExperience.id_field_path), nullable=False
+        String, ForeignKey(PrivacyExperience.id_field_path), nullable=False, index=True
     )
+    experience_config_id = Column(
+        String,
+        ForeignKey(PrivacyExperienceConfig.id_field_path),
+        nullable=True,
+        index=True,
+    )
+
+    experience_config_history_id = Column(
+        String,
+        ForeignKey(PrivacyExperienceConfigHistory.id_field_path),
+        nullable=True,
+        index=True,
+    )
+
+
+def get_privacy_notices_by_region_and_component(
+    db: Session, region: PrivacyNoticeRegion, component: ComponentType
+) -> Query:
+    """
+    Return relevant privacy notices that should be displayed for a certain
+    region in a given component type
+    """
+    return (
+        db.query(PrivacyNotice)
+        .filter(PrivacyNotice.regions.any(region.value))  # type: ignore[attr-defined]
+        .filter(
+            or_(
+                and_(
+                    component == ComponentType.overlay,
+                    PrivacyNotice.displayed_in_overlay,
+                ),
+                and_(
+                    component == ComponentType.privacy_center,
+                    PrivacyNotice.displayed_in_privacy_center,
+                ),
+            )
+        )
+    )
+
+
+def upsert_privacy_experiences_after_notice_update(
+    db: Session, affected_regions: List[PrivacyNoticeRegion]
+) -> Tuple[List[PrivacyExperience], List[PrivacyExperience]]:
+    """
+    Keeps Privacy Experiences in sync with *PrivacyNotices* changes.
+    Create or update PrivacyExperiences based on the PrivacyNotices in the "affected_regions".
+    To be called whenever PrivacyNotices are created or updated (pass in any regions that were potentially affected)
+
+    PrivacyExperiences should not be deleted.  It's okay if no notices are associated with an Experience.
+    """
+    added_experiences: List[PrivacyExperience] = []
+    updated_experiences: List[PrivacyExperience] = []
```

Comment:
This is one of the primary methods added, to keep PrivacyExperiences updated if PrivacyNotices are changed.